HTML: Sequential focus navigation with shadow dom

rakina opened the pull request:
See https://github.com/whatwg/html/pull/4735 for the spec change.

I used the term "flat tree order" in some variable names and comments, but it isn't defined in the spec so it might be confusing. However I'm not sure what else I can use since "composed tree" is also not in the spec, and "shadow-including tree order" is a little bit different because of slots.

rakina committed:
HTML: Sequential focus navigation with shadow dom

domenic commented on html/interaction/focus/sequential-focus-navigation-and-the-tabindex-attribute/focus-tabindex-order-shadow-negative.html:
```diff
@@ -0,0 +1,40 @@
+//      <div #belowSlot tabindex=-1>
+// <div #belowHost tabindex=0>
+
+async_test(async (t) => {
```
This could be `promise_test(` and then you wouldn't need `t.step` or `t.done()`

rakina replied:
Done.

domenic commented on html/interaction/focus/sequential-focus-navigation-and-the-tabindex-attribute/focus-tabindex-order-shadow-negative.html:
```diff
@@ -0,0 +1,40 @@
+  setTabIndex([aboveHost, host, belowHost], 0);
+  resetFocus();
+  const expectedOrder = [aboveHost, host, belowHost]; 
+  for (let el of expectedOrder) {
```
Can this be a utility function too? It seems like the loop is the same in all tests.

```js
await assertFocusOrder([aboveHost, host, belowHost]);
```

rakina replied:
Done.

domenic commented on html/interaction/focus/sequential-focus-navigation-and-the-tabindex-attribute/focus-tabindex-order-shadow-zero-slot-one.html:
```diff
@@ -0,0 +1,74 @@
+  }
+}
+
+function navigateFocusForward() {
```
Can you just use the shared one and then say `const focused = shadowRoot.activeElement || document.activeElement` afterward?

rakina replied:
Done.

domenic reviewed (commented):
Some style nits, but this overall looks good. We may also want to move into shadow-dom/

It would be good to get a second implementers' eye on the setup + expected orders (which are the most important parts of the tests). @rniwa maybe :)

rakina committed:
refactor

rakina commented:
cc @annevk and/or @smaug---- for Firefox?

domenic commented on shadow-dom/focus/resources/shadow-utils.js:
```diff
@@ -69,3 +70,14 @@ function navigateFocusForward() {
   return test_driver.send_keys(document.body, "\ue004");
 }
 
+function assertFocusOrder(expectedOrder) {
```
The `return new Promise` isn't needed here. Just make the whole function async, and then put the loop body inside the function:

```js
async function assertFocusOrder(expectedOrder) {
  let shadowRoot = document.getElementById("host").shadowRoot;
  for (let el of expectedOrder) {
    await navigateFocusForward();
    let focused = shadowRoot.activeElement ? shadowRoot.activeElement : document.activeElement;
    assert_equals(focused, el);
  }
}
```

rakina replied:
Oh awesome, thanks :D

rakina committed:
promise -> async

rniwa commented:
All but one test (focus-tabindex-order-shadow-zero-delegatesFocus.html) with `delegatesFocus` set to true pass in WebKit so these tests look good to me. Perhaps we can land all tests which doesn't involve `delegatesFocus` right now so that we can test coverage on things we ought to have interoperability.

rakina committed:
Remove delegatesFocus test

Will ad the delegatesFocus test in the delegatesFocus PR instead (https://github.com/web-platform-tests/wpt/pull/18035)

rakina commented:
> All but one test (focus-tabindex-order-shadow-zero-delegatesFocus.html) with `delegatesFocus` set to true pass in WebKit so these tests look good to me. Perhaps we can land all tests which doesn't involve `delegatesFocus` right now so that we can test coverage on things we ought to have interoperability.

Yay, thanks for trying it out. I've removed the delegatesFocus test, will move that to https://github.com/web-platform-tests/wpt/pull/18035.
